Add Netatmo temperature services

deosrc opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add Netatmo services for advanced control of room temperature. Allows setting temperature for a given duration, and clearing manually set temperatures to revert to the currently selected schedule.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29883

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Commit:
Update datetime strings to match input_datetime integration

Commit:
Add netatmo service to set temperature

Commit:
Add netatmo service to clear temperature setting

Commit:
Merge remote-tracking branch 'origin/dev' into feature/netatmo-set-temperature-services

Commit:
Merge remote-tracking branch 'origin/dev' into feature/netatmo-set-temperature-services

Commit:
Fix formatting

Commit:
Add tests for new services

Commit:
Fix mypy error

Commit:
Fix formatting

Commit:
Fix formatting

Commit:
Merge remote-tracking branch 'origin/dev' into feature/netatmo-set-temperature-services

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -441,12 +459,31 @@ async def _async_service_set_preset_mode_with_end_datetime(
             end_timestamp,
         )
 
+    async def _async_service_set_temperature(self, **kwargs: Any) -> None:
```
There is already a service for setting temperature so this only makes sense if end_datetime is not optional.

gjohansson-ST replied:
See `_async_service_set_preset_mode_with_end_datetime()` which had the same comments in a previous PR

deosrc replied:
That had confused me a little as to why the end date time was required when it is not for the API. 

I understand wanting to differentiate between `climate.set_temperature` and `netatmo.set_temperature`. It made more sense to me to have it as `set_temperature` as that is what it is primarily doing, just with some additional optional functionality which the netatmo ecosystem provides. This is currently only a single additional argument but that might not always be the case and there may be many combinations of available parameters in future. Having it optional also provides the user with the guidance that it will use the netatmo default if not provided. The set temperature service currently does this silently (3 hours).

I can change it if you still prefer. I certainly don't like it being inconsistent with the set preset service. It just felt like the more sensible name to me for future maintenance and possible expansion.

I've looked through the other PR. As far as I can tell the naming is the only issue so far. I'm not clear on the process so I've marked it as ready for review again for this comment to be considered.

gjohansson-ST commented on homeassistant/components/netatmo/services.yaml:
```diff
@@ -46,6 +46,31 @@ set_preset_mode_with_end_datetime:
       selector:
         datetime:
 
+set_temperature:
```
```suggestion
set_temperature_with_end_datetime:
```

gjohansson-ST reviewed (changes requested):
First pass.
Thinking I will have the same comments here as in the previous PR for presets so good if you take a look there.

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -47,8 +48,10 @@
+    SERVICE_CLEAR_TEMPERATURE_SETTING,
     SERVICE_SET_PRESET_MODE_WITH_END_DATETIME,
     SERVICE_SET_SCHEDULE,
+    SERVICE_SET_TEMPERATURE,
```
```suggestion
    SERVICE_SET_TEMPERATURE_WITH_END_DATETIME,
```

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -143,6 +146,21 @@ def _create_entity(netatmo_device: NetatmoRoom) -> None:
         "_async_service_set_preset_mode_with_end_datetime",
     )
+    platform.async_register_entity_service(
+        SERVICE_SET_TEMPERATURE,
```
```suggestion
        SERVICE_SET_TEMPERATURE_WITH_END_DATETIME,
```

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -143,6 +146,21 @@ def _create_entity(netatmo_device: NetatmoRoom) -> None:
+            vol.Required(ATTR_TARGET_TEMPERATURE): vol.All(
+                vol.Coerce(float), vol.Range(min=7, max=30)
+            ),
+            vol.Optional(ATTR_END_DATETIME): cv.datetime,
```
```suggestion
            vol.Required(ATTR_END_DATETIME): cv.datetime,
```

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -143,6 +146,21 @@ def _create_entity(netatmo_device: NetatmoRoom) -> None:
+            ),
+            vol.Optional(ATTR_END_DATETIME): cv.datetime,
+        },
+        "_async_service_set_temperature",
```
```suggestion
        "_async_service_set_temperature_with_end_datetime",
```

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -441,12 +459,31 @@ async def _async_service_set_preset_mode_with_end_datetime(
 
+    async def _async_service_set_temperature(self, **kwargs: Any) -> None:
+        target_temperature = kwargs[ATTR_TARGET_TEMPERATURE]
+        end_datetime = kwargs.get(ATTR_END_DATETIME)
+        end_timestamp = (
+            int(dt_util.as_timestamp(end_datetime)) if end_datetime else None
+        )
```
```suggestion
        end_datetime = kwargs[ATTR_END_DATETIME]
        end_timestamp = (
            int(dt_util.as_timestamp(end_datetime))
        )
```

gjohansson-ST commented on homeassistant/components/netatmo/const.py:
```diff
@@ -89,12 +89,15 @@
 SERVICE_SET_PERSONS_HOME = "set_persons_home"
 SERVICE_SET_SCHEDULE = "set_schedule"
 SERVICE_SET_PRESET_MODE_WITH_END_DATETIME = "set_preset_mode_with_end_datetime"
+SERVICE_SET_TEMPERATURE = "set_temperature"
```
```suggestion
SERVICE_SET_TEMPERATURE_WITH_END_DATETIME = "set_temperature_with_end_datetime"
```

gjohansson-ST commented on homeassistant/components/netatmo/services.yaml:
```diff
@@ -46,6 +46,31 @@ set_preset_mode_with_end_datetime:
+          min: 7
+          max: 30
+          step: 0.5
+    end_datetime:
+      example: '"2019-04-20 05:04:20"'
+      selector:
+        datetime:
```
```suggestion
    end_datetime:
      required: true
      example: '"2019-04-20 05:04:20"'
      selector:
        datetime:
```

gjohansson-ST reviewed (changes requested):
We should have a clear distinction between the services so to do that we should set this new service as we did with preset e.g. name it with end_datetime and make the field required.

When a user wants to just set a temperature he should use `climate.set_temperature`. When to set with an end datetime use `netatmo.set_temperature_with_end_datetime`.

Did not make changes for the tests, those obviously also needs to be changed with this.

Commit:
Apply suggestions from code review (WIP)

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

Commit:
Complete changes from review suggestions

Commit:
Fix build error

Commit:
Add service to set temperature for time period

deosrc commented:
Ok I've implemented the naming changes as requested.

I also realised my use case (and I expect many other automation use cases) would require the use of jinja to determine the end date & time. It seemed to make sense to have a service to handle time period as well so I've added that.

Commit:
Merge branch 'dev' into feature/netatmo-set-temperature-services

gjohansson-ST commented on homeassistant/components/netatmo/climate.py:
```diff
@@ -441,12 +474,48 @@ async def _async_service_set_preset_mode_with_end_datetime(
+        )
+        await self._room.async_therm_manual(target_temperature, end_timestamp)
+
+    async def _async_service_set_temperature_with_time_period(
```
> I also realised my use case (and I expect many other automation use cases) would require the use of jinja to determine the end date & time. It seemed to make sense to have a service to handle time period as well so I've added that.

I think we should probably try to mimic what's in the "official app/device". Do you select an end time or a run time?
We could of course do both services but I wonder if it really makes sense.

deosrc replied:
For setting the value, the app shows the time period as a slider and defaults to 3h. For displaying the current manual boost, it shows the time it is active until. The API works by timestamp.

I added the extra processing required for time period to avoid having to use as much jinja in automations, but I can see a use case for either. e.g. Set temperature for 3h while clothes are drying or set temperature until 5pm when finishing work.

To set until a given time today, jinja would still be required to work out the date, but it would be much simpler than calculating the time period (e.g. `{{ states('sensor.date') + ' 17:00:00' }}`).

Likewise, having a service to boost for a given timespan would potentially satisfy most use cases in the UI. Working only by datetime would involve some complex jinja to calculate the correct end time.

Maybe HA could handle this automatically in future by having the datetime selector work with times relative to the point when they are triggered, but that is a very different feature.

deosrc replied:
I guess with either one implemented, the alternative can become a blueprint script.

How would you like to proceed?

gjohansson-ST replied:
For preset we implemented it with and end datetime. What's there in the app?

deosrc replied:
For preset you activate it indefinitely, then can "edit" to a specific date and time. There is no time period option for preset in the app.

Display of preset:
![Screenshot_2023-11-20-11-10-56-24_2ac22f0e18cbfbeda85a7e653b55b2c6 (Small)](https://github.com/home-assistant/core/assets/50599779/b12a0fbe-2106-44e8-8c2e-9cb90b8b0965)

Edit preset:
![Screenshot_2023-11-20-11-11-03-08_2ac22f0e18cbfbeda85a7e653b55b2c6 (Small)](https://github.com/home-assistant/core/assets/50599779/7c893350-39c0-4aff-987d-02740d841871)

Edit temperature (select time period but end time also visible):
![Screenshot_2023-11-20-11-11-14-73_2ac22f0e18cbfbeda85a7e653b55b2c6 (Small)](https://github.com/home-assistant/core/assets/50599779/7467335d-4e6f-427f-9d7b-3694c851277b)

Display manual temperature:
![Screenshot_2023-11-20-11-11-19-98_2ac22f0e18cbfbeda85a7e653b55b2c6 (Small)](https://github.com/home-assistant/core/assets/50599779/2f0940fb-e85b-4436-a564-4033c6dfa3d9)

gjohansson-ST replied:
I guess it can make sense to have both then. Thanks

gjohansson-ST commented on tests/components/netatmo/test_climate.py:
```diff
@@ -359,6 +365,174 @@ async def test_service_preset_modes_thermostat(
+    }
+    await simulate_webhook(hass, webhook_id, response)
+
+    assert hass.states.get(climate_entity_livingroom).state == "auto"
```
Does this really test anything as it already was auto?

deosrc replied:
I'm a bit unsure about all of the tests here tbh. This is probably one for @cgtobi to comment on.

The process of the services is:

1. User makes service call
2. Call to API is made through external library
3. Netatmo sends a request to webhook
4. HA updates devices based on webhook

From what I can tell of the tests, only 1, 3 and 4 are being checked.

To compartmentalise the tests, I expect it would be better to have tests to check parts 1 and 2, then separate tests to check parts 3 and 4. I'm not familiar enough with the testing in HA though to know how to do that. It would also increase the scope of this PR by quite a lot to change this.

gjohansson-ST replied:
I mean in the last test just add so it does
1. Make service call to set a temperature with an end datetime
2. Check result
3. Make service call (with new service `async_service_clear_temperature_setting()`) to clear the setting
4. Check result

On the current test we're only asserting the state which is `auto` before and after so actually nothing changes in the test.
It's only doing 3 and 4 right now.

deosrc replied:
Ah I see what you mean. I'll update the test.

Commit:
Expand and fix test

Commit:
Merge branch 'dev' into feature/netatmo-set-temperature-services

gjohansson-ST commented on homeassistant/components/netatmo/strings.json:
```diff
@@ -121,18 +121,50 @@
+          "description": "The target temperature for the device."
+        },
+        "end_datetime": {
+          "name": "End date & time",
```
Please use "links" when the phrases matches up
Example `[%key:component::netatmo::services::xyz::xyz%]`

deosrc replied:
Think that should have sorted it.

Thanks for your patience with this PR. The quick responses have certainly been appreciated.

gjohansson-ST reviewed (changes requested):
Didn't notice before but last bit and then this should be good to go 👍

Commit:
Replace duplicated strings with links

Commit:
Merge remote-tracking branch 'origin/dev' into feature/netatmo-set-temperature-services

gjohansson-ST reviewed (approved):
Thanks @deosrc 👍
